Refactor gradients for extensibility

a-matsuo opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary
closes #9053
closes #9077 

Added `_preprocess` and `_postprocess` into gradients to simplify extensibility.

### Details and comments
- [x] Add `_preprocess` and `_post_process`
  - [x] Generate a circuit with unique parameters from the input
  - [x] Compute the gradient on this "unique-ified" circuit
  - [x] Recombine the results to account for product & chain rules
- [x] Allow a single circuit and observable for inputs. 
- [x] Use `_circuit_key` instead of `id`   
- [x] Make the result of `SamplerGradientResult` sparser.

Cryoris commented on qiskit/algorithms/gradients/base_estimator_gradient.py:
```diff
@@ -166,6 +205,137 @@ def _validate_arguments(
+                        f"{i}-th circuit."
+                    )
+
+    def _preprocess(self, circuits: Sequence[QuantumCircuit]):
```
How about passing the `supported_gates` here as optional argument? That would allow us to remove it from the signature, which would be nice since it's not generally needed in the base.

Cryoris commented on qiskit/algorithms/gradients/base_estimator_gradient.py:
```diff
@@ -45,11 +54,19 @@ def __init__(
+            supported_gates: The list of supported gates to decompose circuits. Defaults to None.
+            add_virtual_var: Whether to add virtual variables to circuits. This is used for the
+                parameter shift gradient. Defaults to False.
+            skip_pre_postprocess: Whether to skip the preprocess and postprocess. Defaults to False.
```
This will need some more information before merging, like explaining that this can be set to true if the circuit has only plain parameters and is in the supported basis. But this can also wait until the end 🙂

Cryoris reviewed (commented):
I started reviewing, but I think it would be super helpful if you could briefly write up what the principle behind the "unique-ification" is as all the variables are a bit confusing 😄 Looking at it I thought we'd only need (1) the unique circuit and (2) a dictionary mapping original parameters to a list of new parameters where the original one occurs. But there seems to be more to it 🙂

a-matsuo commented:
Thank you for the comment! I updated the code to make it easier to understand. Also, I moved some process from the base class to child classes. Let me describe how the code is making a circuit with unique parameters. 
1. Transpile the original circuit using `supported_gates`.
2. Replace parameters in the transpiled circuit with newly created unique parameters.
  If a parameter expression in an instruction is complicated (e.g. when it is not a single parameter), introduce a virtual parameter for `+` and `-pi/2`. 
3. Store the gradient coefficient of each parameter for the chain rule, i.e. `da/d(theta)` in `df/d(theta) = df/da * da/d(theta)`.

Cryoris commented on qiskit/algorithms/gradients/param_shift_estimator_gradient.py:
```diff
@@ -43,7 +41,24 @@ def __init__(self, estimator: BaseEstimator, options: Options | None = None):
         """
-        self._gradient_circuits = {}
+        self._parameter_value_offset_cache = {}
+        self._SUPPORTED_GATES = [
```
This could be defined on class level as it's common to each instance, e.g.
```python
class ...:
  SUPPORTED_GATES = [ ...]
  
  def __init__(...)
```
or if it stays here we should probably choose lowercased variable name 🙂

Cryoris commented on qiskit/algorithms/gradients/param_shift_estimator_gradient.py:
```diff
@@ -43,7 +41,24 @@ def __init__(self, estimator: BaseEstimator, options: Options | None = None):
+        ]
         super().__init__(estimator, options)
 
     def _run(
```
This could be moved to the base class since this is the same for all gradients right? And then only `_run_unique` would need to be abstract 🙂

a-matsuo replied:
But we don't want to apply `preprocess` and `postprocess` for the  finite diff gradient and SPSA gradient (In the future, we might have more numerical gradients and/or approximate gradients). Should we make `_run` like this?

```
def _run():
  if self._prepost_process:
     preprocess()
  
  run_unique
  
  if self._prepost_process:
    postprocess()
_
```

Cryoris reviewed (commented):
The structure looks great, thank you! I left some comments below 🙂

a-matsuo commented on qiskit/algorithms/gradients/base_estimator_gradient.py:
```diff
@@ -166,6 +173,198 @@ def _validate_arguments(
+                    else:
+                        bound_coeff = coeff
+                    original_gradient[i] += bound_coeff * gradient[g_result_indices[g_parameter]]
+            original_gradients.append(original_gradient)
+            original_metadata.append(
```
Maybe, naming is not good. `original_gradient` is the gradient of the original circuit. I wanted to distinguish between  the gradient of the original circuit and the gradient of a circuit with unique parameters.

Cryoris replied:
I see! Yeah maybe explicitly calling the unique gradients `unique_gradients` and the original just `gradients` would be clearer 🙂

a-matsuo commented on qiskit/algorithms/gradients/base_estimator_gradient.py:
```diff
@@ -166,6 +173,198 @@ def _validate_arguments(
+                g_parameters.extend(
+                    g_parameter for g_parameter, _ in gradient_circuit.parameter_map[parameter]
+                )
+        return list(set(g_parameters))
```
It's to keep consistency of signatures of `_run` and `_run_unique` (`parameters` is supposed to be a list).

a-matsuo commented on qiskit/algorithms/gradients/base_estimator_gradient.py:
```diff
@@ -166,6 +173,198 @@ def _validate_arguments(
+                            p: parameter_values_[circuit.parameters.data.index(p)]
+                            for p in coeff.parameters
+                        }
+                        bound_coeff = coeff.bind(local_map)
```
I casted it to float because `bind()` of `ParameterExpression` returns `ParameterExpression`.

Cryoris replied:
Exactly -- I meant that this currently is not cast to float but maybe it _should_ be cast to float 😄

a-matsuo commented on qiskit/algorithms/gradients/base_estimator_gradient.py:
```diff
@@ -166,6 +173,198 @@ def _validate_arguments(
+            )
+        return g_circuits, g_parameter_values, g_parameters
+
+    def _assign_unique_parameters(
```
I moved them into `util.py`.

a-matsuo commented on qiskit/algorithms/gradients/base_estimator_gradient.py:
```diff
@@ -166,6 +173,198 @@ def _validate_arguments(
+                )
+        return list(set(g_parameters))
+
+    def _make_parameter_set(self, circuit: QuantumCircuit, parameters: Sequence[Parameter]):
```
I changed it to `_get_parameter_set`.

Cryoris commented on qiskit/algorithms/gradients/base_estimator_gradient.py:
```diff
@@ -166,6 +173,198 @@ def _validate_arguments(
+            g_parameters_ = self._make_gradient_parameters(circuit, gradient_circuit, parameters_)
+            g_parameter_set = self._make_parameter_set(
+                gradient_circuit.gradient_circuit, g_parameters_
+            )
```
Aha but in that case could this just be `set(g_parameters_)`?

Cryoris commented on qiskit/algorithms/gradients/base_sampler_gradient.py:
```diff
@@ -91,17 +114,121 @@ def _run(
+            # by using the chain rule.
+            gradient = []
+            for parameter in parameter_indices:
+                grad_dist = np.zeros(2**circuit.num_qubits)
```
It would be good to avoid creating this huge vector by default 🙂

a-matsuo replied:
I agree. I updated the code to use dict instead of a large vector.

Cryoris commented on qiskit/algorithms/gradients/base_sampler_gradient.py:
```diff
@@ -91,17 +114,121 @@ def _run(
+        Returns:
+            The results of the gradient of the original circuits.
+        """
+        gradients, metadata = [], []
```
This code is mostly the same as in the estimator, up to the `grad_dist` construction. Is there a way we can avoid the code duplication here? 🤔

a-matsuo replied:
We can move `_preprocess()` and `_postprcess()` in the base sampler gradient and base estimator gradient to `util.py`. Then, we can add some optional argument like `estimator_grad=True` to chose which type of results are expected.

Cryoris commented on qiskit/algorithms/gradients/lin_comb_estimator_gradient.py:
```diff
@@ -73,7 +87,7 @@ def __init__(
                 Higher priority setting overrides lower priority setting.
         """
-        self._gradient_circuits = {}
+        self._lin_comb_cache = {}
```
Why do we need the `gradient_circuits` cache and `lin_comb_cache`? Doesn't the base class already cache the gradient circuits🤔

Cryoris commented on qiskit/algorithms/gradients/lin_comb_sampler_gradient.py:
```diff
@@ -47,80 +71,69 @@ def __init__(self, sampler: BaseSampler, options: Options | None = None):
-            gradients.append(gradient_)
+            gradient = []
+            for dist in result.quasi_dists:
+                grad_dist = np.zeros(2 ** circuits[i].num_qubits)
```
Also here it would be good to avoid the exponentially large distribution per default 🙂

Cryoris commented on qiskit/algorithms/gradients/param_shift_sampler_gradient.py:
```diff
@@ -21,101 +21,89 @@
-            gradients.append(gradient_)
+            gradient = []
+            for dist_plus, dist_minus in zip(result.quasi_dists[:n], result.quasi_dists[n:]):
+                grad_dist = np.zeros(2 ** circuits[i].num_qubits)
```
See above 😄

Cryoris commented on releasenotes/notes/refactor-gradient-d6d315cb256a17db.yaml:
```diff
@@ -0,0 +1,13 @@
+    and ``parameter_values`` and ``parameters`` are also updated to match the circuit.
+    ``_postprocess()`` computes the gradient of the original circuit from the gradient of the
+    gradient circuit by using the chain rule. When users want to implement their own gradient classes,
+    they can use ``_preprocess()`` and ``_postprocess()``. Then, they implement ``run_unique()``
```
```suggestion
    they can use ``_preprocess()`` and ``_postprocess()``. Then, they implement ``_run_unique()``
```

Cryoris commented on releasenotes/notes/refactor-gradient-d6d315cb256a17db.yaml:
```diff
@@ -0,0 +1,13 @@
+    gradient circuit by using the chain rule. When users want to implement their own gradient classes,
+    they can use ``_preprocess()`` and ``_postprocess()``. Then, they implement ``run_unique()``
+    that calculates a gradient of a circuit with unique parameters.
+  - |
```
Could we post that under a `fixes:` part instead of feature? 😄

ElePT commented on qiskit/algorithms/gradients/param_shift_estimator_gradient.py:
```diff
@@ -17,100 +17,91 @@
 
-        # combine the results
         try:
             results = [job.result() for job in jobs]
```
While in principle this line enables async execution of the jobs, which should be good for performance, in practice we have found that when dealing with the runtime primitives, it is significantly faster to batch all circuits together and submit a single primitive job than sending multiple jobs. @mrossinek and I opened this draft PR (#9077) with the old gradient design. And I think it would be good to have in all of the gradient classes after this refactoring.

ElePT replied:
Here is my suggestion to apply this batching, but maybe you can come up with a better design:

```
    def _run_unique(
        self,
        circuits: Sequence[QuantumCircuit],
        observables: Sequence[BaseOperator | PauliSumOp],
        parameter_values: Sequence[Sequence[float]],
        parameter_sets: Sequence[set[Parameter]],
        **options,
    ) -> EstimatorGradientResult:
        """Compute the estimator gradients on the given circuits."""
        jobs, metadata = [], []
        job_circuits, job_observables, job_param_values = [], [], []
        all_n = []
        for circuit, observable, parameter_values_, parameter_set in zip(
            circuits, observables, parameter_values, parameter_sets
        ):
            metadata.append({"parameters": [p for p in circuit.parameters if p in parameter_set]})
            # Make parameter values for the parameter shift rule
            param_shift_parameter_values = _make_param_shift_parameter_values(
                circuit, parameter_values_, parameter_set
            )
            n = len(param_shift_parameter_values)

            # Combine inputs to the estimator
            job_circuits.extend([circuit] * n)
            job_observables.extend([observable] * n)
            job_param_values.extend(param_shift_parameter_values)
            all_n.append(n)

        # Send single job with all circuits
        jobs = self._estimator.run(
            job_circuits,
            job_observables,
            job_param_values,
            **options,
        )
        try:
            results = jobs.result()
        except Exception as exc:
            raise AlgorithmError("Estimator job failed.") from exc

        # Compute the gradients
        gradients = []
        partial_sum_n = 0
        for n in all_n:
            result = results.values[partial_sum_n: partial_sum_n + n]
            gradient_ = (result[:n//2] - result[n//2:]) / 2
            gradients.append(gradient_)
            partial_sum_n += n

        opt = self._get_local_options(options)
        return EstimatorGradientResult(gradients=gradients, metadata=metadata, options=opt)
```

ElePT replied:
Something that might be important to note is that by bundling together these circuits, we might be pushing the limits of how many circuits the primitives can handle, but my take here is that this is not something we should worry about, but should be handled by the actual primitives (but still good to keep in mind in case we see strange behaviours in the future).

ElePT reviewed (commented):
Thanks Matsuo-san! I left a suggestion for the gradient implementation, please take a look at my comment and feel free to contact me if you see anything that doesn't make sense.

Cryoris commented on qiskit/algorithms/gradients/finite_diff_sampler_gradient.py:
```diff
@@ -84,74 +85,121 @@ def _run(
+            partial_sum_n += n
+            gradients.append(gradient)
+
+        # for result in results:
```
This should be removed 🙂

Cryoris reviewed (approved):
LGTM, thanks a lot for the work -- this greatly simplifies the implementation of other gradients like for the fidelity or classical gradients!

Cryoris commented:
I'm adding the `Bugfix` and `New Feature` labels to make sure we don't forget this in the changelog. However I'm not 100% sure if this actually qualifies as feature, since it's not a new functionality we've added but simplifies extensions of the gradients. But I guess better label it and then we can discuss whether to include it 🙂
